### PR TITLE
AWS: replace aws copy for encryption with encrypted EBS volumes

### PR DESCRIPTION
**Pull request by abhinavdahiya**
xref: https://issues.redhat.com/browse/CORS-1352
depends on https://github.com/openshift/cluster-api-provider-aws/pull/303

Amazon supports using [encrypted root EBS volumes][1]. So instead of copying the AMI and encrypting in the process we can use the encypted EBS volumes instead which saves us making the AMI copy during install.

The change also allows the user to provide KMS key to the terraform instead of the default KMS key.

Using this change the instance boots with public RHCOS AMI, and the EBS volume is encrypted with default KMS key for the account and region.
```console
[4:42:15] ➜  installer git:(ami_no_copy) ✗ AWS_PROFILE=openshift-dev aws ec2 describe-volumes --region ca-central-1 --volume-ids vol-09ceb58bc9ac93602
{
    "Volumes": [
        {
            "Attachments": [
                {
                    "AttachTime": "2020-03-13T23:30:09.000Z",
                    "Device": "/dev/xvda",
                    "InstanceId": "i-0bdec7b64f2b9a607",
                    "State": "attached",
                    "VolumeId": "vol-09ceb58bc9ac93602",
                    "DeleteOnTermination": true
                }
            ],
            "AvailabilityZone": "ca-central-1b",
            "CreateTime": "2020-03-13T23:30:09.789Z",
            "Encrypted": true,
            "KmsKeyId": "arn:aws:kms:ca-central-1:<REDACTED>:key/1ffbfa61-7250-42de-9a75-cac64a8f0e4d",
            "Size": 120,
            "SnapshotId": "snap-0c8421bbe06f5f69d",
            "State": "in-use",
            "VolumeId": "vol-09ceb58bc9ac93602",
            "Iops": 360,
            "Tags": [
                {
                    "Key": "kubernetes.io/cluster/adahiya-1-djk4q",
                    "Value": "owned"
                },
                {
                    "Key": "Name",
                    "Value": "adahiya-1-djk4q-master-1-vol"
                }
            ],
            "VolumeType": "gp2"
        }
    ]
}
[4:42:28] ➜  installer git:(ami_no_copy) ✗ AWS_PROFILE=openshift-dev aws ec2 describe-instances --region ca-central-1 --instance-ids i-0bdec7b64f2b9a607
{
    "Reservations": [
        {
            "Groups": [],
            "Instances": [
                {
                    "AmiLaunchIndex": 0,
                    "ImageId": "ami-0d235b4920e9def96",
                    "InstanceId": "i-0bdec7b64f2b9a607",
                    "InstanceType": "m4.xlarge",
                    "LaunchTime": "2020-03-13T23:30:09.000Z",
...
                    "BlockDeviceMappings": [
                        {
                            "DeviceName": "/dev/xvda",
                            "Ebs": {
                                "AttachTime": "2020-03-13T23:30:09.000Z",
                                "DeleteOnTermination": true,
                                "Status": "attached",
                                "VolumeId": "vol-09ceb58bc9ac93602"
                            }
                        }
                    ],
...
                    "Tags": [
                        {
                            "Key": "Name",
                            "Value": "adahiya-1-djk4q-master-1"
                        },
                        {
                            "Key": "kubernetes.io/cluster/adahiya-1-djk4q",
                            "Value": "owned"
                        }
                    ],
...
        }
    ]
}
[4:44:31] ➜  installer git:(ami_no_copy) ✗
```

Users can provide the KMS Key ARN which should be used to encrypt the EBS volumes otherwise the default KMS key for the region will be used.

[1]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
tried to bring the cluster up using cluster-bot using cmd: `/msg @clusterbot launch openshift/cluster-api-provider-aws#303,openshift/installer#3293 aws`

**Comment by abhinavdahiya** [comment]
> We have measured the impact of this change and decided that it does not improve the situation

From quick testing, this reduces the 4-5 minutes _on less traffic account_ and maybe a lot more in _ci_ account as the AMI copy is completely removed.
But from the testing also saw that even if the control-plane booted very quickly, they were stuck waiting for internal DNS zone and records, _around 2 minutes_, to be created to reach out to the machine-config-server for Ingition.

> OR We have implemented this change and it proven that it reduces the overhead associated with copying and encrypting our boot images

I think not copying does reduce any overhead

> Understand how non default KMS keys can be used for encryption.

non-default KMS keys can easily be provided to the install-config per machine pool i.e. separate for control-plane and compute which is maybe nice.

> Are there any limitations?

- from [encrypted EBS volume][1] docs, they list `Supported Instance Types`. These seems to include all the known instance classes at least the m4 and m5 so we should be good.

- Still figuring out if this is supported in non-public regions like US Gov Cloud, China

[1]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html

**Comment by cgwalters** [comment]
We clearly can't ever delete bootimages that has ever been pinned by a released OpenShift version.

I'd also hesitate at deleting bootimages that were *ever* pinned (e.g. during a devel cycle).

Beyond that, I don't see a reason for us not to prune the bootimages eventually.

So this should clearly be fine.

**Comment by abhinavdahiya** [comment]
/hold cancel
deaa20d63c43fc598dcbc3211c145537ad6b7d0b beings in the correct vendor bump for cluster-apir-provider-aws

File: docs/user/aws/customization.md [hunk 1]
```diff
@@ -20,6 +20,8 @@ Beyond the [platform-agnostic `install-config.yaml` properties](../customization
         This is only valid for `type` `io1`.
     * `size` (optional integer): Size of the root volume in gibibytes (GiB).
     * `type` (optional string):  The [type of volume][volume-type].
+    * `kmsKeyARN` (optional string): The [ARN of KMS key][ebs-kms-key] that should be used to encrypt the EBS volume.
+        Example ARN values are: `arn:aws:kms:us-east-1:111122223333:key/1234abcd-12ab-34cd-56ef-1234567890ab` or `arn:aws:kms:us-east-1:111122223333:alias/my-key`
```
Comment (wking) [inline docs/user/aws/customization.md]: mention the default KMS key here too?  Seem more important in this customer-ish facing docs than in the dev-only `variables-aws.tf` ;).

File: data/data/aws/main.tf [hunk 2]
```diff
@@ -18,7 +18,7 @@ provider "aws" {
 module "bootstrap" {
   source = "./bootstrap"
 
-  ami                      = aws_ami_copy.main.id
+  ami                      = var.aws_ami
```
Comment (wking) [inline data/data/aws/main.tf]: Adjust the bootstrap module to always encrypt with the default KMS key?  Or share in `var.aws_master_root_volume_kms_key_id`?

Comment (abhinavdahiya) [reply data/data/aws/main.tf]: fixed in 9ab21a975fb5b5f36d32b05258be36068ddf9467
the bootstrap host is always encrypted like before, but uses default or user provided KMS key.

File: docs/user/aws/customization.md (hunk 1 above)
Comment (abhinavdahiya) [reply docs/user/aws/customization.md]: fixed in 17030b3bdb997b8d5c64ec19b48e94da12f1be84

**Pull request merged**